x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -437,55 +476,106 @@ public void getAuthenticationAndMetadata(String token, ActionListener<Tuple<Auth
+            logger.warn("failed to get access token [{}] because index [{}] is not available", tokenId, tokensIndex.aliasName());
             listener.onFailure(frozenTokensIndex.getUnavailableReason());
         } else {
-            final GetRequest getRequest = client.prepareGet(tokensIndex.aliasName(), getTokenDocumentId(userTokenId)).request();
-            final Consumer<Exception> onFailure = ex -> listener.onFailure(traceLog("get token from id", userTokenId, ex));
+            final GetRequest getRequest = client.prepareGet(tokensIndex.aliasName(), getTokenDocumentId(tokenId)).request();
+            final Consumer<Exception> onFailure = ex -> listener.onFailure(traceLog("get token from id", tokenId, ex));
             tokensIndex.checkIndexVersionThenExecute(
-                ex -> listener.onFailure(traceLog("prepare tokens index [" + tokensIndex.aliasName() + "]", userTokenId, ex)),
+                ex -> listener.onFailure(traceLog("prepare tokens index [" + tokensIndex.aliasName() + "]", tokenId, ex)),
                 () -> executeAsyncWithOrigin(
                     client.threadPool().getThreadContext(),
                     SECURITY_ORIGIN,
                     getRequest,
                     ActionListener.<GetResponse>wrap(response -> {
                         if (response.isExists()) {
-                            @SuppressWarnings("unchecked")
                             Map<String, Object> accessTokenSource = (Map<String, Object>) response.getSource().get("access_token");
                             if (accessTokenSource == null) {
                                 onFailure.accept(new IllegalStateException("token document is missing the access_token field"));
                             } else if (accessTokenSource.containsKey("user_token") == false) {
                                 onFailure.accept(new IllegalStateException("token document is missing the user_token field"));
-                            } else {
-                                @SuppressWarnings("unchecked")
-                                Map<String, Object> userTokenSource = (Map<String, Object>) accessTokenSource.get("user_token");
-                                listener.onResponse(UserToken.fromSourceMap(userTokenSource));
-                            }
+                            } else if (tokenVersion.onOrAfter(VERSION_GET_TOKEN_DOC_FOR_REFRESH)
+                                && accessTokenSource.containsKey("token") == false) {
+                                    onFailure.accept(new IllegalStateException("token document is missing the user_token.token field"));
+                                } else if (tokenVersion.onOrAfter(VERSION_GET_TOKEN_DOC_FOR_REFRESH)
+                                    && response.getSource().get("refresh_token") != null
+                                    && ((Map<String, Object>) response.getSource().get("refresh_token")).containsKey("token") == false) {
```

Comment:
I know the source parsing is a madness, but it's only done in this one place. I'm not sure doing some parsing in a constructor for `AccessAndRefreshTokens` would help.
However, I've pushed [03f6084](https://github.com/elastic/elasticsearch/pull/97395/commits/03f60845dbb3b9497bc5999ef448a465fca5ca6c) to address the parsing madness. It removes some of the nesting, to avoid going over the line length limit which results in dubious formatting.
Let me know if this is looking more reasonable now.

Reply:
Looks much better. 👍